### PR TITLE
[REG-1939] capture logs during recording

### PR DESCRIPTION
**Pull request by abeizer**
Creates a `LoggingObserver` component that stores received logs in a queue. Each tick, logs from the observer are dequeued and written to a `/logs/{tickNumber}.jsonl` file. When recording is stopped, the `/logs` dir is zipped and uploaded to RGService alongside our other recording resources.

image

!! Note: The log file will be empty if no logs were observed leading up to the given tick

Sample of logs from RGBossRoom:
```jsonl
{"timestamp":"2024-08-29T20:02:46.7976702+00:00","level":"Log","message":"ConnectionManager: Changed connection state from OfflineState to StartingHostState.","stackTrace":"UnityEngine.Debug:Log (object)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:ChangeState (Unity.BossRoom.ConnectionManagement.ConnectionState) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:114)\nUUnity.BossRoom.ConnectionManagement.OfflineState:StartHostIP (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionState/OfflineState.cs:56)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:StartHostIp (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:167)\nUnity.BossRoom.Gameplay.UI.IPUIMediator:HostIPRequest (string,string) (at Assets/Scripts/Gameplay/UI/IPUIMediator.cs:86)\nUnity.BossRoom.Gameplay.UI.IPHostingUI:OnCreateClick () (at Assets/Scripts/Gameplay/UI/IPHostingUI.cs:37)\nUnityEngine.EventSystems.EventSystem:Update () (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/EventSystem.cs:514)\n"}
{"timestamp":"2024-08-29T20:02:46.7996703+00:00","level":"Log","message":"Created relay allocation with join code ","stackTrace":"UnityEngine.Debug:Log (object)\nUnity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10:MoveNext () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:80)\nSystem.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10> (Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10&)\nUnity.BossRoom.ConnectionManagement.StartingHostState:StartHost ()\nUnity.BossRoom.ConnectionManagement.StartingHostState:Enter () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:31)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:ChangeState (Unity.BossRoom.ConnectionManagement.ConnectionState) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:122)\nUUnity.BossRoom.ConnectionManagement.OfflineState:StartHostIP (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionState/OfflineState.cs:56)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:StartHostIp (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:167)\nUnity.BossRoom.Gameplay.UI.IPUIMediator:HostIPRequest (string,string) (at Assets/Scripts/Gameplay/UI/IPUIMediator.cs:86)\nUnity.BossRoom.Gameplay.UI.IPHostingUI:OnCreateClick () (at Assets/Scripts/Gameplay/UI/IPHostingUI.cs:37)\nUnityEngine.EventSystems.EventSystem:Update () (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/EventSystem.cs:514)\n"}
{"timestamp":"2024-08-29T20:02:46.8208931+00:00","level":"Warning","message":"[Netcode] Runtime Network Prefabs was not empty at initialization time. Network Prefab registrations made before initialization will be replaced by NetworkPrefabsList.","stackTrace":"UnityEngine.Debug:LogWarning (object)\nUnity.Netcode.NetworkLog:LogWarning (string) (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Logging/NetworkLog.cs:28)\nUnity.Netcode.NetworkPrefabs:Initialize (bool) (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Configuration/NetworkPrefabs.cs:71)\nUnity.Netcode.NetworkConfig:InitializePrefabs () (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Configuration/NetworkConfig.cs:273)\nUnity.Netcode.NetworkManager:Initialize (bool) (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Core/NetworkManager.cs:787)\nUnity.Netcode.NetworkManager:StartHost () (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Core/NetworkManager.cs:930)\nUnity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10:MoveNext () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:83)\nSystem.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10> (Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10&)\nUnity.BossRoom.ConnectionManagement.StartingHostState:StartHost ()\nUnity.BossRoom.ConnectionManagement.StartingHostState:Enter () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:31)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:ChangeState (Unity.BossRoom.ConnectionManagement.ConnectionState) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:122)\nUUnity.BossRoom.ConnectionManagement.OfflineState:StartHostIP (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionState/OfflineState.cs:56)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:StartHostIp (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:167)\nUnity.BossRoom.Gameplay.UI.IPUIMediator:HostIPRequest (string,string) (at Assets/Scripts/Gameplay/UI/IPUIMediator.cs:86)\nUnity.BossRoom.Gameplay.UI.IPHostingUI:OnCreateClick () (at Assets/Scripts/Gameplay/UI/IPHostingUI.cs:37)\nUnityEngine.EventSystems.EventSystem:Update () (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/EventSystem.cs:514)\n"}
{"timestamp":"2024-08-29T20:02:46.8812189+00:00","level":"Log","message":"ConnectionManager: Changed connection state from StartingHostState to HostingState.","stackTrace":"UnityEngine.Debug:Log (object)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:ChangeState (Unity.BossRoom.ConnectionManagement.ConnectionState) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:114)\nUnity.BossRoom.ConnectionManagement.StartingHostState:OnServerStarted () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:53)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:OnServerStarted () (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:137)\nUnity.Netcode.NetworkManager:StartHost () (at Library/PackageCache/com.unity.netcode.gameobjects@1.3.1/Runtime/Core/NetworkManager.cs:997)\nUnity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10:MoveNext () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:83)\nSystem.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10> (Unity.BossRoom.ConnectionManagement.StartingHostState/<StartHost>d__10&)\nUnity.BossRoom.ConnectionManagement.StartingHostState:StartHost ()\nUnity.BossRoom.ConnectionManagement.StartingHostState:Enter () (at Assets/Scripts/ConnectionManagement/ConnectionState/StartingHostState.cs:31)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:ChangeState (Unity.BossRoom.ConnectionManagement.ConnectionState) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:122)\nUUnity.BossRoom.ConnectionManagement.OfflineState:StartHostIP (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionState/OfflineState.cs:56)\nUnity.BossRoom.ConnectionManagement.ConnectionManager:StartHostIp (string,string,int) (at Assets/Scripts/ConnectionManagement/ConnectionManager.cs:167)\nUnity.BossRoom.Gameplay.UI.IPUIMediator:HostIPRequest (string,string) (at Assets/Scripts/Gameplay/UI/IPUIMediator.cs:86)\nUnity.BossRoom.Gameplay.UI.IPHostingUI:OnCreateClick () (at Assets/Scripts/Gameplay/UI/IPHostingUI.cs:37)\nUnityEngine.EventSystems.EventSystem:Update () (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/EventSystem.cs:514)\n"}
```

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs [hunk 1]
```diff
@@ -20,6 +21,32 @@
 // ReSharper disable once LoopCanBeConvertedToQuery - Better performance using indexing vs enumerators
 namespace RegressionGames.StateRecorder
 {
+    
+    class TickDataToWriteToDisk
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs:25]: I see now why this was added, lots of data now. However, I wonder if this is faster as a class or a struct, my C# knowledge is not as strong there, maybe @RG-nAmKcAz would know? It might be faster to create a struct than allocate a new object? https://schwabencode.com/blog/2021/09/20/dotnet-class-vs-struct-performance

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs [hunk 2]
```diff
@@ -814,5 +899,33 @@ private void RecordBotSegment(string directoryPath, long tickNumber, byte[] json
...
+                        RGDebug.LogException(nextTask.Exception, $"ERROR: Unable to record logs for tick # {tickNumber}");
+                    }
+
+                });
+                _fileWriteTasks.Add((path,fileWriteTask));
+            }
+            catch (Exception e)
+            {
+                RGDebug.LogException(e, $"ERROR: Unable to record logs for tick # {tickNumber}");
+            }
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs:928]: Do we need to worry about ticks attempting to write to this file at the same time? And additionally, logs from ticks being added in the incorrect order? I am honestly not sure

**Review by vontell — CHANGES_REQUESTED** [review]
Looking good! Just a few comments about clarifications and performance, as well as a potential race condition.

File: src/gg.regression.unity.bots/Runtime/Prefabs/RGOverlayCanvasV2.prefab
[the remark this answers is not in the document]
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Prefabs/RGOverlayCanvasV2.prefab:1]: I added the LoggingObserver to the overaly

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs [hunk 3]
```diff
@@ -143,6 +174,17 @@ private async Task HandleEndRecording(long tickCount, DateTime startTime, DateTi
...
 
+            var zipTask4 = Task.Run(() =>
+            {
+                // Then save the logs separately
+                RGDebug.LogInfo($"Zipping logs recording replay to file: {logsPathPrefix}.zip");
+                using (ZipArchive zip = ZipFile.Open(logsPathPrefix + ".zip", ZipArchiveMode.Create))
+                {
+                    zip.CreateEntryFromFile(logsPathPrefix + ".jsonl", "logs.jsonl");
+                }
+                RGDebug.LogInfo($"Finished zipping replay to file: {logsPathPrefix}.zip");
```
[the remark this answers is not in the document]
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs:185]: Correct on all counts

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs (hunk 2 above)
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs:928]: Ooh I'm not sure either. Need to read up on how tasks work with I/O...

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs [hunk 4]
```diff
@@ -723,11 +807,12 @@ private void ProcessTicks()
             }
         }
 
-        private void ProcessTick(string directoryPath, long tickNumber, byte[] botSegmentJson, byte[] jsonData, int tickScreenshotWidth, int tickScreenshotHeight, byte[] tickScreenshotData)
+        private void ProcessTick(TickDataToWriteToDisk tickData)
         {
-            RecordBotSegment(directoryPath, tickNumber, botSegmentJson);
-            RecordJson(directoryPath, tickNumber, jsonData);
-            RecordJPG(directoryPath, tickNumber, tickScreenshotWidth, tickScreenshotHeight, tickScreenshotData);
+            RecordBotSegment(tickData.directoryPrefix, tickData.tickNumber, tickData.botSegmentJson);
```
[the remark this answers is not in the document]
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs]: updated

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs (hunk 2 above)
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs]: This approach definitely seems unsafe. What's interesting is that Ashley's original implementation didn't account for this issue, unless I'm missing something from that PR. Maybe our logs were so sparse that the Tasks completed quickly enough that we never hit a case of concurrent / out-of-order write operations? 

Either way, I've pushed a change that moves ownership of I/O to the `LoggingObserver` itself. Since the timestamp is recorded as logs come in, we shouldn't need to write to file every single tick - this will now wait for the previous write task to complete before queuing a new one. This works in BossRoom, but I haven't gotten to try this out with <partner project> yet (it produces so many logs that it should be great for testing)

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs (hunk 1 above)
Comment (abeizer) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs]: Yeah we're passing so many variables between multiple methods that adding a new one requires several line changes. I assumed we were passing individual args rather than an object/struct to restrict who can modify values, but if @RG-nAmKcAz's intent is to optimize performance then I'll remove this and accept that we'll just have long method signatures

Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs]: I'll leave this open for Zack to give a final say, but LGTM!

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs (hunk 2 above)
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs]: Also leaving this open so Zack has some context on the idea and can give a second pair of eyes on this potential issue.

**Review by vontell — APPROVED** [review]

**Comment by vontell** [comment]
All good on my end!

**Review by addisonbgross — CHANGES_REQUESTED** [review]
Can you please add your changes in `RGOverlayCanvasV2` to the `RGOverlayCanvas` prefab I've been working on for the Bot Sequence stuff

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/LoggingObserver.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        
+        /// <summary>
+        /// Called each tick to flush any available logs to disk.
+        /// We may not actually write logs to file every tick:
+        /// If there's a running task accessing the destination file, then we'll skip this tick and try again next tick.
+        /// </summary>
+        public void FlushLogs(string filePath, CancellationToken token)
+        {
+            if (_fileWriteTask != null && !_fileWriteTask.IsCompleted)
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/LoggingObserver.cs:72]: The internals of this method need to be wrapped in a `lock` statement.

If this method were to be called 2x close in time, it can start 2 file write tasks to the same file in parallel.  The reason for this is that both calls can get past this first check thinking the current fileWriteTask is completed because there is no locking to ensure one of them created a new task before the other one checks.

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]
Small concurrency issue

**Comment by abeizer** [comment]
@RG-nAmKcAz @addisonbgross @vontell re-requesting review: this now outputs logs in separate files per tick. description updated

**Review by addisonbgross — APPROVED** [review]
Providing everyone else's concerns have been met, I like this

**Review by RG-nAmKcAz — APPROVED** [review]
